[BEAM-2500] Add S3 FileSystem to Java SDK

jacobmarble opened the pull request:
My first contribution, will submit ICLA later today.

Ran `mvn clean verify` against sdks/java/io only, because I'm currently in a place with limited power and bandwidth.

There is some technical discussion in the JIRA.
https://issues.apache.org/jira/browse/BEAM-2500

Please notice that there is one persistent thread pool, and that copy operations can create a new thread pool per copy. Happy to hear feedback and discuss.

iemejia commented:
Retest this please

iemejia commented:
Great work Jacob, The build did not pass for some simple errors on findbugs, you can validate this locally with `mvn clean verify -Prelease`.

iemejia commented:
R: @iemejia (I add myself more from a end-user validation point of view and some Beam IO experience).
R: @lukecwik (Sorry to bother you but since you implemented most of HDFS and know BeamFS really well you seemd to me like a perfect candidate, feel free to minus if you are busy and can't do the review).

lukecwik commented:
I'll take a look.

jacobmarble commented:
I should ask, is an ICLA needed for a patch this size?

lukecwik commented:
I believe so since your adding a whole feature.

jacobmarble commented:
Submitted ICLA to Apache.

lukecwik commented on sdks/java/io/amazon-web-services/pom.xml:
```diff
@@ -0,0 +1,154 @@
+  <packaging>jar</packaging>
+
+  <properties>
+    <aws-java-sdk.version>1.11.194</aws-java-sdk.version>
```
We have other dependencies on AWS libraries at version 1.11.18

We should match that here or upgrade our existing usage to have the same version.

jacobmarble replied:
Upgraded Kinesis version.

lukecwik commented on sdks/java/io/amazon-web-services/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-sdks-java-core</artifactId>
+      <scope>compile</scope>
```
I believe setting the compile scope is redundant

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/options/S3Options.java:
```diff
@@ -0,0 +1,57 @@
+  void setS3StorageClass(String value);
+
+  @Description("Size of S3 upload chunks; max upload object size is this value multiplied by 10000")
+  @Nullable
```
Whats the default?

jacobmarble replied:
Added to the flag description

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/options/S3Options.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Options used to configure Amazon Web Services S3.
+ */
+public interface S3Options extends AwsOptions {
```
nit: It looks cleaner if get/set pair methods have no newline separators.

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3ResourceId.java:
```diff
@@ -0,0 +1,153 @@
+/**
+ * {@link ResourceId} implementation for Amazon Web Services S3.
+ */
+public class S3ResourceId implements ResourceId {
```
Add a test for S3ResourceId which just runs ResourceIdTester.runResourceIdBattery(...)

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/pom.xml:
```diff
@@ -0,0 +1,154 @@
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>provided</scope>
```
I believe this should be test scope.

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java:
```diff
@@ -0,0 +1,573 @@
+/**
+ * S3Util wraps the Amazon Web Services client library.
+ */
+public class S3Util {
```
It looks like these methods are only called from S3FileSystem class. Please merge these methods into that class instead of exposing this class.

jacobmarble replied:
Consolidated S3Util into S3FileSystem

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3WritableByteChannel.java:
```diff
@@ -0,0 +1,145 @@
+/**
+ * A writable S3 object, as a {@link WritableByteChannel}.
+ */
+class S3WritableByteChannel implements WritableByteChannel {
```
Move this out of util and into org.apache.beam.sdk.io.aws.s3

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3ReadableSeekableByteChannel.java:
```diff
@@ -0,0 +1,154 @@
+/**
+ * A readable S3 object, as a {@link SeekableByteChannel}.
+ */
+class S3ReadableSeekableByteChannel implements SeekableByteChannel {
```
Move this out of util and into org.apache.beam.sdk.io.aws.s3

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3Path.java:
```diff
@@ -0,0 +1,569 @@
+/**
+ * Implements the Java NIO {@link Path} API for AWS S3 paths.
+ */
+public class S3Path implements Path, Serializable {
```
Once everything that access this class is inside org.apache.beam.sdk.io.aws.s3, please make this class package private

jacobmarble replied:
Consolidated S3Path into S3ResourceId and made S3ResourceId package private

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3Path.java:
```diff
@@ -0,0 +1,569 @@
+    // - "Labels" are defined as bucket name parts separated by single period delimiter.
+    // - Each label starts and ends with lowercase letter or number.
+    // - Bucket name cannot be formatted as an IP address.
+    if (bucket == null) {
```
Is there a library supplied by Amazon that does this validation for us?

jacobmarble replied:
I'm not sure if it's in the SDK or an error returned by the AWS API, but either way, it's unnecessary and I reduced this to null-and-empty checks.

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3Path.java:
```diff
@@ -0,0 +1,569 @@
+  /**
+   * A coder for {@link S3Path}.
+   */
+  public static class Coder extends SerializableCoder<S3Path> {
```
I'd strongly recommend against extending a concrete class. S3Path already implements Serializable, so this class should just be unnecessary.

jacobmarble replied:
Consolidated S3Path into S3ResourceId, and eliminated the coder. Users specify paths as strings, so this shouldn't have been necessary.

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java:
```diff
@@ -0,0 +1,573 @@
+  }
+
+  @Override
+  protected void finalize() {
```
Hm why not just make this AutoCloseable?

jacobmarble replied:
This is now folded into S3Filesystem.

Looking at what other FileSystems do:
GcsFileSystem creates a thread pool per batch of work
HadoopFileSystem and LocalFileSystem don't use concurrency

So I'll switch to the GcsFileSystem approach.

jacobmarble replied:
Done

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java:
```diff
@@ -0,0 +1,573 @@
+
+  private void delete(String bucket, Collection<String> keys) throws IOException {
+    // S3 SDK: "You may specify up to 1000 keys."
+    checkArgument(keys.size() <= 1000, "only 1000 keys can be deleted per request");
```
Is this a limitation that propagates into Beam, or do you have chunking into 1000-key batches somewhere at a higher level?

jacobmarble replied:
The method is private, and the only usage is a few lines away. I'll change this to a constant value to make this more readable and less error-prone.

jacobmarble replied:
Done

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java:
```diff
@@ -0,0 +1,573 @@
+
+  private static MatchResult.Metadata createBeamMetadata(S3Path path, long sizeBytes) {
+    // TODO: Address https://issues.apache.org/jira/browse/BEAM-1494
+    // It is incorrect to set IsReadSeekEfficient true for files with content encoding set to gzip.
```
It would be good to address this in the current PR, or at least fail on files whose encoding is gzip, because silently skipping this issue can lead to pretty dramatic performance if not addressed.

jacobmarble replied:
TODO

jacobmarble replied:
Done

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/util/S3Util.java:
```diff
@@ -0,0 +1,573 @@
+   * @return a string with the regular expression this glob expands to
+   */
+  @VisibleForTesting
+  static String wildcardToRegexp(String globExp) {
```
It feels like we already have a function like this somewhere

jacobmarble replied:
It's from GcsFileSystem. Happy to consolidate, but being new to the project, I don't know what that would look like. Do you have a suggestion?

jacobmarble replied:
TODO

jacobmarble replied:
@jkff please take a moment to answer this question

jkff commented on sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/util/MatchResultMatcher.java:
```diff
@@ -0,0 +1,118 @@
+ * Hamcrest {@link Matcher} to match {@link MatchResult}. Necessary because {@link
+ * MatchResult#metadata()} throws an exception under normal circumstances.
+ */
+public class MatchResultMatcher extends BaseMatcher<MatchResult> {
```
Looks like something potentially useful for other filesystems.

jacobmarble replied:
I had that in mind when I wrote it. What do you think about this?
test/o/a/b/sdk/io/testing

jacobmarble replied:
TODO

jacobmarble replied:
@jkff another question here

jkff commented on sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/util/S3UtilTest.java:
```diff
@@ -0,0 +1,367 @@
+import org.mockito.Mockito;
+
+/**
+ * Test case for {@link S3Util}.
```
Suggest to verify that there is near-100% code coverage in this PR, at least for the non-exception paths.

jacobmarble replied:
TODO

jkff commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3Path.java:
```diff
@@ -0,0 +1,569 @@
+/*
```
What role does S3Path play in this PR? Beam only requires that you provide a ResourceId, so I suppose S3Path is an internal implementation detail? If so - are all of its functions actually used by the implementation?

jacobmarble replied:
The PR was originally based on the GCS implementation, which I now see has more complexity (and unused code) than needed. The author probably had something in mind besides just implementing a FileSystem.

I think you'll see that today's changes simplify things. IMHO this should be squashed before merging, and I'm happy to do that whenever.

jacobmarble replied:
Done

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
SUCCESS 
 
--none--

jkff commented:
Thanks! Luke, could you do the next round of review?

jacobmarble commented:
TODO: describe testing as requested by Eugene

asfgit commented:
FAILURE
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
SUCCESS 
 
--none--

asfgit commented:
FAILURE
 
--none--

asfgit commented:
FAILURE
 
--none--

lukecwik commented:
The code structure from GcsFileSystem is quite poor because it has evolved from GcsUtil which was created before Dataflow 1.0 release. During the migration to "file systems", there were plans to remove GcsUtil and clean up the implementation of GcsFileSystem but they were never acted upon.

I'll take another look now that a bunch of stuff has been moved around.

jacobmarble commented:
@lukecwik The current build failure is caused by the AWS update in Kinesis, shouldn't be hard to resolve. If you plan to review again in the next 24h then I'll hold off pushing more commits for the moment.

lukecwik commented:
Yeah, I'm reviewing right now. Also long as you don't squash your commits adding changes won't complicate my review.

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3ReadableSeekableByteChannel.java:
```diff
@@ -0,0 +1,153 @@
+    this.amazonS3 = checkNotNull(amazonS3, "amazonS3");
+    this.path = checkNotNull(path, "path");
+    try {
+      contentLength = amazonS3.getObjectMetadata(path.getBucket(), path.getKey())
```
Either fix or add future TODO with JIRA link:
Store size within S3ResourceId as Optional<Long> meaning that if its populated you don't need to do this call here.
You get the size for free from match and the most common pattern will be match + read.

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3ReadableSeekableByteChannel.java:
```diff
@@ -0,0 +1,153 @@
+      throw new ClosedChannelException();
+    }
+    if (!destinationBuffer.hasArray()) {
+      throw new UnsupportedOperationException("ByteBuffer.hasArray() must be true");
```
It doesn't seem like your restricted to in memory byte buffers only based upon the code below, do you really need to check that hasArray is true?

jacobmarble replied:
Nope. Done.

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3ReadableSeekableByteChannel.java:
```diff
@@ -0,0 +1,153 @@
+    }
+    checkArgument(newPosition >= 0, "newPosition too low");
+    checkArgument(newPosition < contentLength, "new position too high");
+    if (s3Object != null) {
```
if newPosition == position, just return instead of closing and resetting the work that has been done so far.

jacobmarble replied:
Agreed. Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3WritableByteChannel.java:
```diff
@@ -0,0 +1,144 @@
+    if (!isOpen()) {
+      throw new ClosedChannelException();
+    }
+    if (!sourceBuffer.hasArray()) {
```
It doesn't seem like you need to have hasArray to be true since none of the code calls array() directly.

jacobmarble replied:
Done

lukecwik commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3WritableByteChannel.java:
```diff
@@ -0,0 +1,144 @@
+    this.amazonS3 = checkNotNull(amazonS3, "amazonS3");
+    this.path = checkNotNull(path, "path");
+    checkArgument(uploadBufferSizeBytes > 0, "uploadBufferSizeBytes");
+    this.uploadStream = new ByteArrayOutputStream(uploadBufferSizeBytes);
```
This only controls the initial buffer size of the BAOS and not the max size so you won't honor the maximum if the person sent you a giant byte buffer to write.

jacobmarble replied:
Done

lukecwik commented:
What you have done here is great, was wondering if you had time to follow up on the review comments?

jacobmarble commented:
Hey Luke, I got 90% done with your latest over the long weekend, should have time to wrap up tomorrow after work, so you should see something Wednesday morning.

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/options/AwsOptions.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public interface AwsOptions extends PipelineOptions {
+
+  @Description("AWS access key ID")
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/options/AwsOptions.java:
```diff
@@ -0,0 +1,40 @@
+  void setAwsAccessKeyId(String value);
+
+  @Description("AWS secret access key")
+  String getAwsSecretAccessKey();
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+  private static final int DEFAULT_UPLOAD_BUFFER_SIZE_BYTES =
+      Runtime.getRuntime().maxMemory() < 512 * 1024 * 1024
+          ? MINIMUM_UPLOAD_BUFFER_SIZE_BYTES
+          : 64 * 1024 * 1024;
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+
+  @Override
+  protected List<MatchResult> match(List<String> specs) throws IOException {
+    List<S3ResourceId> paths =
```
Ack

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+  @Override
+  protected WritableByteChannel create(S3ResourceId resourceId, CreateOptions createOptions)
+      throws IOException {
+    return new S3WritableByteChannel(amazonS3, resourceId, storageClass, s3UploadBufferSizeBytes);
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+          amazonS3.initiateMultipartUpload(initiateUploadRequest);
+      uploadId = initiateUploadResult.getUploadId();
+
+      ObjectMetadata objectMetadata =
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+      // Don't pollute the main thread pool, which this was called from.
+      // Instead, create a small thread pool just for this copy operation.
+      int threadQuantity = Math.min(MAX_THREADS_PER_CONCURRENT_COPY, tasks.size());
+      executorService = Executors.newFixedThreadPool(threadQuantity);
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+   */
+  private <T> List<T> callTasks(Collection<Callable<T>> tasks) throws IOException {
+
+    ListeningExecutorService executorService =
```
Done

jacobmarble commented on sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java:
```diff
@@ -0,0 +1,599 @@
+      throw new IOException("executor service was interrupted");
+
+    } finally {
+      executorService.shutdown();
```
Done

jacobmarble commented:
There are a couple of TODOs remaining. Here are some notes from tonight:
- rebased to master and squashed
- WordCount works, reading from and writing to S3
- need to do some bigger testing
- haven't looked at coverage yet
- Kinesis tests fail for me both before and after this PR, so I made a fix and we'll see what Jenkins does with it. If that doesn't work, AWS SDK version 1.11.24 is the last version that should pass without any change in Beam.
- there are two open questions to @jkff , mentioned inline
- tried to do more tonight, but fixing the content encoding bug turned out to be more complex than I anticipated.

jacobmarble commented:
Coverage notes:

S3FileSystem - 80% LOC
S3ResourceId - 93% LOC
missing: exception handling, copy(List, List), rename, matchNewResource

copy(List, List) and rename is two LOC: copy, delete
matchNewResource doesn't do a lot

jacobmarble commented:
Practical testing notes:

Wrote a TextIO.Read => TextIO.Write pipeline to test lots of files and data.

Used Dataflow runner, Oregon region
GS and S3 regions are both Oregon
Workers are n1-standard-4
Autoscaling enabled, 100 workers max
No compression

Pipeline 1: GS to S3
~3000 files read from GS
3095 files written to S3
1274848854 lines
368107865893 bytes
47 workers at peak
time 11m16s

Pipeline 2: GS to GS
same input as pipeline 1
same lines, same bytes
38 workers at peak
time 5m0s

Pipeline 3: S3 to GS
input is output of pipeline 1
3150 files written to GS
same lines, same bytes
38 workers at peak
time 5m14s

jacobmarble commented:
Luke, the AWS credentials solution you proposed, if I understand it correctly, doesn't work on Dataflow. [DefaultAWSCredentialsProviderChain](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/index.html?com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html) wants to get the credentials from one of the following (copy-paste from Slack):

- environment variables (don't know how to set these on a Dataflow worker)
- Java system properties (I haven't used this)
- credentials file at `~/.aws/credentials` (pretty sure the Beam SDK doesn't support this)

I'll do one more load test this morning, with gzip compression.

jacobmarble commented:
I just tried to run a pipeline that includes this package, but doesn't use it. Looks like the S3FileSystemRegistrar creates an instance of S3FileSystem even if I don't use it, which means the S3FileSystem constructor probably shouldn't create a thread pool or enforce flag requirements. Now the GcsUtil class makes sense.

lukecwik commented:
Have the S3FileSystemRegistrar not return any filesystems to register if the required flags are missing. You can print a debug statement as to why S3 support is not being enabled.

Having GcsUtil created within PipelineOptions by GcsFileSystem has little to do with not starting threads or flag enforcement and has more to do with no one spending the time to get rid of GcsUtil and migrate it into GcsFilesystem.

lukecwik commented:
The [DefaultAWSCredentialsProviderChain](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/index.html?com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html) is the best default to use for Apache Beam.

For Dataflow users wanting to access AWS services, they should really set one of the static credentials providers like:
* [ClasspathPropertiesFileCredentialsProvider](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/ClasspathPropertiesFileCredentialsProvider.html)
* [AWSStaticCredentialsProvider](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/AWSStaticCredentialsProvider.html)
* or use one of the dynamic ones and set the correct properties.

It is important to document that users need to use a credentials provider that is compatible with the remote execution location and we should make sure serializing PipelineOptions with a few common credentials providers works.

Once we are using the portability framework then the user will be able to build a container containing any files (~/.aws/credentials), additional flags to the JVM (system properties), or additional processes AWS credentials provider.

jacobmarble commented:
We're probably close to done writing code, just this auth question remains, yes?

Is this what you have in mind when you say "they should really set one of the static credentials providers" ? I tried the following snippet, and it works with the DirectRunner, but fails with the DataflowRunner.

```
interface ExampleOptions extends PipelineOptions, S3Options {

    String getSourceS3Filename();
    void setSourceS3Filename(String value);

    String getDestinationS3FilenamePrefix();
    void setDestinationS3FilenamePrefix(String value);
  }

  public static void main(String[] args) {
    PipelineOptionsFactory.register(ExampleOptions.class);
    ExampleOptions options =
        PipelineOptionsFactory.fromArgs(args).create().as(ExampleOptions.class);
    AWSCredentials awsCredentials =
        new BasicAWSCredentials("test key", "test secret");
    options.setAwsCredentialsProvider(new AWSStaticCredentialsProvider(awsCredentials));
    Pipeline pipeline = Pipeline.create(options);

    pipeline.apply(TextIO.read().from(options.getSourceS3Filename()))
        .apply(TextIO.write().to(options.getDestinationS3FilenamePrefix()).withSuffix(".txt"));

    pipeline.run().waitUntilFinish();
  }
```

The Dataflow error:

```
com.amazonaws.SdkClientException: Unable to load AWS credentials from any provider in the chain
        at com.amazonaws.auth.AWSCredentialsProviderChain.getCredentials(AWSCredentialsProviderChain.java:131)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.getCredentialsFromContext(AmazonHttpClient.java:1163)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.runBeforeRequestHandlers(AmazonHttpClient.java:762)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:724)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
```

You can check out the failing AWS client code here:
https://github.com/aws/aws-sdk-java/blob/7844c64cf248aed889811bf2e871ad6b276a89ca/aws-java-sdk-core/src/main/java/com/amazonaws/auth/AWSCredentialsProviderChain.java#L131

Because of this, I suspect that my snippet doesn't properly "set one of the static credential providers". Clearly the DirectRunner and DataflowRunner handle my example differently.

I'll try to reply promptly to this thread for the rest of the week. Hopefully we can resolve this last issue.

jacobmarble commented:
I made a small change to allow building with, but not using, the package. Tested by building with this package as a dependency, then reading and writing GCS without --awsRegion and without any valid AWS credentials present.

lukecwik commented:
The reason why its not working is that the DirectRunner doesn't serialize/deserialize PipelineOptions so the `@JsonIgnore` doesn't come into effect on the `getAwsCredentialsProvider`. This is where my comment about being able to serialize PipelineOptions with a few well known CredentialsProviders would allow us to ship this across to the worker on runners with remote execution like Dataflow/Spark/Flink/...

Filed https://issues.apache.org/jira/browse/BEAM-3344 for DirectRunner to serialize/deserialize PipelineOptions before execution.

jkff commented:
What's the next action on this PR? Does BEAM-3344 need to be fixed before this PR can proceed, or does that issue merely expose a symptom of a different problem? If so, do we know how to fix it?

lukecwik commented:
BEAM-3344 exposed a different problem and is a pretty easy fix to for serialization/deserialization of pipeline options before execution. Fixing BEAM-3344 does not block this change nor does this change block fixing BEAM-3344.

The next steps on this PR is for a couple of implementations of common `AwsCredentialsProvider` Jackson serialization modules so that the credential provider can be round tripped to/from workers for Flink/Spark/Dataflow/...

johnmac-ibm commented:
@jacobmarble How is `--awsCredentialsProvider` intended to be used? I'm perhaps missing something but I was not able to point it to my own custom provider.

lukecwik commented:
Once there is a JSON -> AwsCredentialsProvider mapping it will be expected that --awsCredentialsProvider={JSON for AwsCredentialsProvider} can be used. My last comment to @jacobmarble was to add support for these mappings for the common AwsCredentialsProvider types (or find a library that does this already.) Since the mapping doesn't yet exist, its hard to say what the mapping looks like.

jacobmarble commented:
Hi, just to let you all know that I'm still around. Work has me busy with other projects right now, so it's harder to find time for this last fix to the PR.

iemejia commented:
@johnmac-ibm Excellent idea (the serviceEndpoint one) also to debug with a docker emulator like localstack. We recently added this also for kinesis so we can try to make it homogeneous once we unify all amazon-web-services IOs in the future.

jacobmarble commented:
I got AWSCredentialsProvider serialization figured out, but there's a small problem due to the way that I setup polymorphism. Getting closer, though.

lukecwik commented:
Thats good news.

jacobmarble commented:
Tested last commit (auth de/serialization) by running a simple job (80 million records, GCS to S3 and back) on the DataflowRunner with 10 workers and the following values for the awsCredentialsProvider flag:

- `'{"@type":"AWSStaticCredentialsProvider","awsAccessKeyId":"redacted","awsSecretKey":"redacted"}'`
  - Success

- `'{"@type":"ThisIsGarbage","awsAccessKeyId":"redacted","awsSecretKey":"redacted"}'`
  - Failure: java.io.IOException: AWS credential provider type 'ThisIsGarbage' is not supported

I believe most implementations of AWSCredentialsProvider are supported, but not all. There are a bunch of providers that are singletons, so they were easy to add. Unit tests confirm they should work, but I don't have a way to test them. The more complex providers are certainly within reach for someone that needs them (and can test them).

The default provider is DefaultAWSCredentialsProviderChain, which should work for a lot of development environments using DirectRunner.

I did not perform bigger tests with this auth commit because they are expensive to run and because they wouldn't prove anything about auth working or not.

Thanks to @lukecwik for an extra hand in getting through the credentials serialization.

There have been a handful of requests for various features to be included in this PR. The PR has taken months already; can we get this initial implementation submitted with the current feature set? That will unlock anyone else from submitting their own, smaller PRs to improve the S3 filesystem. For example, I have a Redshift CL that I would really like to share.

lukecwik commented:
I think this is worthy to get in even though it doesn't have gradle build rules defined.

iemejia commented:
Great to see this one finally merged, congrats @jacobmarble and thanks for the great contribution !

iemejia commented:
@johnmac-ibm Created https://issues.apache.org/jira/browse/BEAM-3550 to track the support of the serviceEndpoint, feel free to take it or tell me your Beam JIRA id and I will assign it to you.

jacobmarble commented:
Thanks @iemejia , it feels good!

johnmac-ibm commented:
@iemejia Thanks, I'm johnmac-ibm on JIRA as well.

iemejia commented:
I just assigned the JIRA to you @johnmac-ibm.

hostirosti commented:
@jack749 Do you know how writing to multiple buckets in different regions can be configured? Details about the issue I'm running into: https://issues.apache.org/jira/browse/BEAM-11257#

What I'd like to do is something like this:
```
public void processElement(ProcessContext c) {

AwsOptions options = c.getPipelineOptions().as(AwsOptions.class);

... 

options.setAwsCredentialsProvider(<bucket01_auth>);

options.setAwsRegion(<bucket01_region>);

...

ResourceId resourceId = FileSystems.matchNewResource(<bucket01_filename>, false);
try (WritableByteChannel out = FileSystems.create(resourceId, MimeTypes.TEXT))

{ out.write(contents); }
...

options.setAwsCredentialsProvider(<bucket02_auth>);

options.setAwsRegion(<bucket02_region>);

ResourceId resourceId = FileSystems.matchNewResource(<bucket02_filename>, false);
try (WritableByteChannel out = FileSystems.create(resourceId, MimeTypes.TEXT)) { out.write(contents); }

}
```

hostirosti commented:
Solution seems to be to use 

`FileSystems.setDefaultPipelineOptions(options);`

at runtime before using FileSystems.create(..)
